extensions/eclipse-che-theia-prefs-provider/src/browser/prefs-provider.ts

```
@@ -0,0 +1,130 @@
+            .reduce((result: [string, string][], preferences: { [key: string]: string }) => {
+                Object.keys(preferences).forEach(key => {
+                    result.push(<[string, string]>[key, preferences[key]]);
+                });
+                return result;
+            }, []);
+    }
+
+    private async setPluginProperties(props: [string, string][]): Promise<void> {
+        await this.workspaceService.roots;
+        const workspace = this.workspaceService.workspace;
+        if (!workspace) {
+            throw new Error('Failed to get Theia workspace.');
+        }
+        for (const [key, value] of props) {
+            await this.preferenceService.set(key, value, PreferenceScope.Workspace, workspace.uri);
+        }
+    }
+
+    private async getWorkspace(): Promise<che.workspace.Workspace> {
+        let workspaceId;
+        workspaceId = await this.envVariablesServer.getValue('CHE_WORKSPACE_ID');
+
+        if (!workspaceId || !workspaceId.value) {
+            throw new Error('Failed to get current workspace ID.');
+        }
+
+        const remoteApi = await this.restClient();
+        return await remoteApi.getById<che.workspace.Workspace>(await workspaceId.value);
+    }
+
+    private async restClient(): Promise<IRemoteAPI> {
+        const cheApiEndpoint = await this.envVariablesServer.getValue('CHE_API_EXTERNAL');
+
+        if (!cheApiEndpoint || !cheApiEndpoint.value) {
+            throw new Error('Failed to get Eclipse Che API endpoint.');
+        }
+
+        return WorkspaceClient.getRestApi({
+            baseUrl: await cheApiEndpoint.value
+        });
+    }
+
+    async restorePluginProperties(): Promise<void> {
+        try {
+            const workspace = await this.getWorkspace();
+            const propsTuples = this.getPluginsProperties(workspace);
+            return this.setPluginProperties(propsTuples);
+        } catch (e) {
+            throw new Error('Preferences Provider Extension: ' + e.message);
```

Why not just let the exception propagate here? This loses all information about where the error occurred.

---

Good point, thanks, I'll fix